README.md
# `hhi` An R Package to Calculate and Visualize the Herfindahl-Hirschman Index
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/hhi)](http://cran.r-project.org/package=hhi)
[![DOI](http://joss.theoj.org/papers/10.21105/joss.00828/status.svg)](https://doi.org/10.21105/joss.00828)
[![Downloads](http://cranlogs.r-pkg.org/badges/grand-total/hhi)](http://cranlogs.r-pkg.org/)

## Why should I use `hhi`?

The Herfindahl-Hirschman Index (HHI) is a widely used measure of concentration in a variety of fields including, business, economics, political science, finance, and others. Though intuitively simple to calculate (summed squared market shares of firms/actors in a single space), calculation of HHI can get onerous, especially as the number of firms/actors increases.

Thus, the `hhi` package calculates the concentration of a market/space based on a vector of values corresponding with market shares of individual firms/actors. The packages is intended for simple, intuitive usage requiring the name of the data frame and the name of the vector (variable) corresponding with the market shares in quotation marks, placed directly in the function. Calling the function `hhi` will generate the HHI measure based on the values supplied to the function.

Then, to visualize a time series of HHI values, there is a simple plotting function included, `plot_hhi`. Usage is similarly straightforward, requiring the name of the data frame, the name of the variable corresponding with the time indicator in quotation marks, and then the name of the variable corresponding with the market shares also in quotation marks. Leveraging `ggplot2`, the result is a visual rendering of the supplied vector of HHI values over the specified range of time. The function supports any measure of time, such as, years, quarters, months, etc.

Note that `plot_hhi` is a relatively inflexible function meant for quick visual rendering of a vector of HHI values over a period of time. Users are advised to generate original plots with other plotting functions and packages beyond `hhi` to allow for greater flexibility in customizing visual output according to specific needs.

## How do I use `hhi`?

Below are examples of the pair of functions in the package, including `hhi` to generate the measure, and then `plot_hhi` to plot a vector of HHI values over a specified period of time.

```{R}

install.packages("hhi")
library(hhi)

# Create some simple data
a <- c(1,2,3,4) # firm id
b <- c(20,30,40,10) # market share of each firm (should total 100% of market share)
x <- data.frame(a,b) # df

# Call hhi to calculate the HHI based on the vector of market share values in object b in data frame x
hhi(x, "b")

# Plot a time series of fictitious HHI values
hhi <- c(45,60,50,100,94,15,88,200,215,68,
47,62,52,102,96,17,90,202,217,70) # create a vector of HHI values by year
year <-c(1990,1991,1992,1993,1994,1995,1996,1997,1998,1999,
2000,2001,2002,2003,2004,2005,2006,2007,2008,2009) # create a vector of corresponding years
data <- data.frame(hhi, year) # combine into a data frame

# Now plot the HHI by year over time using the "plot_hhi" command
plot_hhi(data, "year", "hhi")

```

## How do I get `hhi `?

You can download the package from [CRAN](https://CRAN.R-project.org/package=hhi). If you have any questions or find any bugs requiring fixing, feel free to open an issue ticket, pull request, or simply [contact me](http://www.philipdwaggoner.com/) if you would prefer.

Please cite the JOSS paper associated with the package using the TeX entry:

@article{hhi2018,<br/>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;title = {The hhi Package: Streamlined Calculation and Visualization of Herfindahl-Hirschman Index Scores},<br/>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;author = {Philip D. Waggoner},<br/>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;journal={Journal of Open Source Software},<br/>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;year = {2018},<br/>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;volume={3},<br/>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;number={28, 828},<br/>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;note={DOI: 10.21105/joss.00828}<br/>
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;}

## References
1. Herfindahl, Orris C. 1950. "Concentration in the steel industry." Ph.D. dissertation, Columbia University.
2. Hirschman, Albert O. 1945. "National power and structure of foreign trade." Berkeley, CA: University of California Press.
3. Rhoades, Stephen A. 1993. "The herfindahl-hirschman index." Federal Reserve Bulletin 79: 188.
4. Waggoner, Philip D. 2018. "The hhi Package: Streamlined Calculation and Visualization of Herfindahl-Hirschman Index Scores." Journal of Open Source Software, 3(28), 828.

Thanks and enjoy!


NEWS
---
title: "NEWS.md"
author: "Philip D. Waggoner"
date: "8/16/2018"
output: html_document
---

# `hhi` 1.2.0
## Calculate and Visualize the Herfindahl-Hirschman Index

### Changes

* Added the Journal of Open Source Software citation for the paper associated with this package.

* Streamlined a portion of the code for quicker and more efficient running.

* Cleaned up the description.

* Updated author affiliation.

## How do I get `hhi `?

The package is released on CRAN. If you have any questions or find any bugs requiring fixing, please feel free to contact me either directly (see the `DESCRIPTION` file for more) or by opening a ticket on GitHub. Thanks and enjoy!


tests/testthat.R
library(testthat)
library(hhi)

test_check("hhi")


tests/testthat/hhi_test.R
context("hhi tests")

test_that("function operates properly", {
  data(x)
  d <- x[ ,s]
  shares <- try(sum(x[ ,s]))
  expect_true(is.data.frame(x), label = "data.frame returned")
  expect_true(is.numeric(d), label = "numeric values returned")
  expect_warning(shares < 100 | shares > 100)
  expect_error(any(d < 0))
})

test_that("hhi returns sum of squared shares", {
  expect_equal(hhi("x", "s"), sum((x[ ,s])^2))
})


R/hhi.R
#' Calculates the Herfindahl-Hirschman Index
#'
#' Computes the Herfindahl-Hirschman Index of a market/space, which is a measure of concentration, based on the share size of all individual firms/actors
#'@usage hhi(x, "s")
#'@param x Name of the data frame
#'@param s Name of the vector (variable) from the data frame, x, corresponding with stored market share values
#'@details Allows for placing the objects directly from working datasets (including, data frame and the market share variable name in quotes) into the function for intuitive usage.
#'@return hhi A measure of market concentration
#'@note The vector of "share" values should be comprised of integers corresponding to total share of individual firms/actors (e.g., df$s <- c(35, 40, 5, 10, 6, 4) # 6 firms totaling 100 percent of the market). The measure is often used as a measure of competition, where 0 is perfect competitiveness and 10,000 is a perfect monopoly.
#'@references Hirschman, Albert O. 1945. "National power and structure of foreign trade." Berkeley, CA: University of California Press.
#'@references Herfindahl, Orris Clemens. 1950. "Concentration in the steel industry." Ph.D. dissertation, Columbia University.
#'@references Rhoades, Stephen A. 1993. "The herfindahl-hirschman index." Federal Reserve Bulletin 79: 188.
#'@examples
#' a <- c(1,2,3,4) # arbitrary firm id
#' b <- c(20,30,40,10) # market share of each firm
#' x <- data.frame(a,b) # create data frame
#' hhi(x, "b") # calculate market concentration based on firms' share sizes
#'@export
hhi <- function(x, s){
  if(!is.data.frame(x)) {
    stop('"x" must be data frame\n',
         'You have provided an object of class: ', class(x)[1])
  }
  shares <- try(sum(x[ ,s]))
  if(shares < 100 | shares > 100) warning('shares, "s", do not sum to 100')
  d <- x[ ,s]
  if(!is.numeric(d)) {
    stop('"s" must be numeric vector\n',
         'You have provided an object of class: ', class(d)[1])
  }
  if (any(d < 0)) {
    stop('vector "s" must contain only positive values')
}
hhi <- sum(d^2)
  
  return(hhi)
}


old ex/footwear example.R
# First, install the "hhi" package, then load the library
install.packages("hhi")
library(hhi)

# Next, read in Data: Footwear .txt from 2012 to 2017 (Source: © Euromonitor International)
footwear <- read.table(".../footwear.txt") # for the .txt file, download "footwear.txt"

# Now, calculate HHI for years 2016 and then 2017
hhi(footwear, "ms.2016")
hhi(footwear, "ms.2017")

# Now, calculate and store HHI for each year in the data set (2012-2017)
hhi.12 <- hhi(footwear, "ms.2012")
hhi.13 <- hhi(footwear, "ms.2013")
hhi.14 <- hhi(footwear, "ms.2014")
hhi.15 <- hhi(footwear, "ms.2015")
hhi.16 <- hhi(footwear, "ms.2016")
hhi.17 <- hhi(footwear, "ms.2017")

# Combine and create df for plot
hhi <- rbind(hhi.12, hhi.13, hhi.14, hhi.15, hhi.16, hhi.17)
year <- c(2012, 2013, 2014, 2015, 2016, 2017)
hhi.data <- data.frame(year, hhi)

# Finally, generate HHI time series plot using the "plot_hhi" command
plot_hhi(hhi.data, "year", "hhi")


R/plot_hhi.R
## In Roxygen syntax -- update accordingly

#' Plots a Herfindahl-Hirschman Index Time Series
#'
#' Generates a plot of a vector of Herfindahl-Hirschman Index (HHI) values over a period of time
#'@usage plot_hhi(x, "t", "h")
#'@param x Name of the data frame
#'@param t Name of the vector (variable) from the data frame, x, corresponding with a measure of time (year, month, etc.)
#'@param h Name of the vector (variable) from the data frame, x, corresponding with the HHI values
#'@details Allows for placing objects directly from working datasets (including, data frame, measure of time, and stored HHI values in quotes) into the function for intuitive usage. Function supports any measure of time.
#'@return hhi.plot A plot of HHI values over time
#'@note The "plot_hhi" function is meant for quick visual rendering of a vector of HHI values over a period of time leveraging ggplot2. Users are advised to generate original plots with other plotting packages beyond "hhi" to allow for greater flexibility in customizing visual output according to specific needs.
#'@references Hirschman, Albert O. 1945. "National power and structure of foreign trade." Berkeley, CA: University of California Press.
#'@references Herfindahl, Orris Clemens. 1950. "Concentration in the steel industry." Ph.D. dissertation, Columbia University.
#'@references Rhoades, Stephen A. 1993. "The herfindahl-hirschman index." Federal Reserve Bulletin 79: 188.
#'@examples
#'hhi <- c(45,60,50,100,94,15,88,200,215,68,
#'47,62,52,102,96,17,90,202,217,70)
#'year <- c(1990,1991,1992,1993,1994,1995,1996,1997,1998,1999,
#'2000,2001,2002,2003,2004,2005,2006,2007,2008,2009)
#'data <- data.frame(hhi, year)
#'plot_hhi(data, "year", "hhi")
#'@export
plot_hhi <- function(x,t,h){
  if (!requireNamespace("ggplot2", quietly = TRUE)) {
    stop("Package \"ggplot2\" needed for this function to work. Please install it.",
         call. = FALSE)
  }

  t <- x[ ,t]
  h <- x[ ,h]

  hhi.plot <- ggplot2::ggplot(x, ggplot2::aes(t, h)) +
    ggplot2::geom_point() +
    ggplot2::geom_line() +
    ggplot2::xlab("Time") +
    ggplot2::ylab("Herfindahl-Hirschman Index") +
    ggplot2::ggtitle("Herfindahl-Hirschman Index Over Time") +
    ggplot2::theme_bw()

  return(hhi.plot)

}
